// llamaindex/dist/cjs/storage/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    PostgresDocumentStore: function() {
        return _PostgresDocumentStore.PostgresDocumentStore;
    },
    PostgresIndexStore: function() {
        return _PostgresIndexStore.PostgresIndexStore;
    },
    PostgresKVStore: function() {
        return _PostgresKVStore.PostgresKVStore;
    },
    SimpleDocumentStore: function() {
        return _SimpleDocumentStore.SimpleDocumentStore;
    },
    SimpleIndexStore: function() {
        return _SimpleIndexStore.SimpleIndexStore;
    },
    SimpleKVStore: function() {
        return _SimpleKVStore.SimpleKVStore;
    }
});
_export_star(require("@llamaindex/core/storage/chat-store"), exports);
const _PostgresDocumentStore = require("./docStore/PostgresDocumentStore.js");
const _SimpleDocumentStore = require("./docStore/SimpleDocumentStore.js");
_export_star(require("./docStore/types.js"), exports);
_export_star(require("./FileSystem.js"), exports);
const _PostgresIndexStore = require("./indexStore/PostgresIndexStore.js");
const _SimpleIndexStore = require("./indexStore/SimpleIndexStore.js");
_export_star(require("./indexStore/types.js"), exports);
const _PostgresKVStore = require("./kvStore/PostgresKVStore.js");
const _SimpleKVStore = require("./kvStore/SimpleKVStore.js");
_export_star(require("./kvStore/types.js"), exports);
_export_star(require("./StorageContext.js"), exports);
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
