// @mistralai/mistralai/sdk/sdk.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mistral = void 0;
const sdks_js_1 = require("../lib/sdks.js");
const agents_js_1 = require("./agents.js");
const batch_js_1 = require("./batch.js");
const chat_js_1 = require("./chat.js");
const classifiers_js_1 = require("./classifiers.js");
const embeddings_js_1 = require("./embeddings.js");
const files_js_1 = require("./files.js");
const fim_js_1 = require("./fim.js");
const finetuning_js_1 = require("./finetuning.js");
const models_js_1 = require("./models.js");
class Mistral extends sdks_js_1.ClientSDK {
    get models() {
        var _a;
        return ((_a = this._models) !== null && _a !== void 0 ? _a : (this._models = new models_js_1.Models(this._options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new files_js_1.Files(this._options)));
    }
    get fineTuning() {
        var _a;
        return ((_a = this._fineTuning) !== null && _a !== void 0 ? _a : (this._fineTuning = new finetuning_js_1.FineTuning(this._options)));
    }
    get batch() {
        var _a;
        return ((_a = this._batch) !== null && _a !== void 0 ? _a : (this._batch = new batch_js_1.Batch(this._options)));
    }
    get chat() {
        var _a;
        return ((_a = this._chat) !== null && _a !== void 0 ? _a : (this._chat = new chat_js_1.Chat(this._options)));
    }
    get fim() {
        var _a;
        return ((_a = this._fim) !== null && _a !== void 0 ? _a : (this._fim = new fim_js_1.Fim(this._options)));
    }
    get agents() {
        var _a;
        return ((_a = this._agents) !== null && _a !== void 0 ? _a : (this._agents = new agents_js_1.Agents(this._options)));
    }
    get embeddings() {
        var _a;
        return ((_a = this._embeddings) !== null && _a !== void 0 ? _a : (this._embeddings = new embeddings_js_1.Embeddings(this._options)));
    }
    get classifiers() {
        var _a;
        return ((_a = this._classifiers) !== null && _a !== void 0 ? _a : (this._classifiers = new classifiers_js_1.Classifiers(this._options)));
    }
}
exports.Mistral = Mistral;
