// @datastax/astra-db-ts/dist/data-api/index.js
"use strict";
// Copyright Datastax, Inc
// SPDX-License-Identifier: Apache-2.0
// noinspection DuplicatedCode
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAPIHttpError = exports.CollectionNotFoundError = exports.UpdateManyError = exports.CursorIsStartedError = exports.BulkWriteError = exports.InsertManyError = exports.TooManyDocumentsToCountError = exports.DeleteManyError = exports.CollectionAlreadyExistsError = exports.DataAPIError = exports.CumulativeDataAPIError = exports.DataAPIResponseError = exports.DataAPITimeoutError = exports.Db = exports.FindCursor = exports.Collection = void 0;
var collection_1 = require("./collection");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return collection_1.Collection; } });
var cursor_1 = require("./cursor");
Object.defineProperty(exports, "FindCursor", { enumerable: true, get: function () { return cursor_1.FindCursor; } });
var db_1 = require("./db");
Object.defineProperty(exports, "Db", { enumerable: true, get: function () { return db_1.Db; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "DataAPITimeoutError", { enumerable: true, get: function () { return errors_1.DataAPITimeoutError; } });
Object.defineProperty(exports, "DataAPIResponseError", { enumerable: true, get: function () { return errors_1.DataAPIResponseError; } });
Object.defineProperty(exports, "CumulativeDataAPIError", { enumerable: true, get: function () { return errors_1.CumulativeDataAPIError; } });
Object.defineProperty(exports, "DataAPIError", { enumerable: true, get: function () { return errors_1.DataAPIError; } });
Object.defineProperty(exports, "CollectionAlreadyExistsError", { enumerable: true, get: function () { return errors_1.CollectionAlreadyExistsError; } });
Object.defineProperty(exports, "DeleteManyError", { enumerable: true, get: function () { return errors_1.DeleteManyError; } });
Object.defineProperty(exports, "TooManyDocumentsToCountError", { enumerable: true, get: function () { return errors_1.TooManyDocumentsToCountError; } });
Object.defineProperty(exports, "InsertManyError", { enumerable: true, get: function () { return errors_1.InsertManyError; } });
Object.defineProperty(exports, "BulkWriteError", { enumerable: true, get: function () { return errors_1.BulkWriteError; } });
Object.defineProperty(exports, "CursorIsStartedError", { enumerable: true, get: function () { return errors_1.CursorIsStartedError; } });
Object.defineProperty(exports, "UpdateManyError", { enumerable: true, get: function () { return errors_1.UpdateManyError; } });
Object.defineProperty(exports, "CollectionNotFoundError", { enumerable: true, get: function () { return errors_1.CollectionNotFoundError; } });
Object.defineProperty(exports, "DataAPIHttpError", { enumerable: true, get: function () { return errors_1.DataAPIHttpError; } });
__exportStar(require("./events"), exports);
__exportStar(require("./ids"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./embedding-providers"), exports);
