// llamaindex/dist/cjs/storage/docStore/PostgresDocumentStore.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PostgresDocumentStore", {
    enumerable: true,
    get: function() {
        return PostgresDocumentStore;
    }
});
const _global = require("@llamaindex/core/global");
const _PostgresKVStore = require("../kvStore/PostgresKVStore.js");
const _KVDocumentStore = require("./KVDocumentStore.js");
const _utils = require("./utils.js");
const DEFAULT_TABLE_NAME = "llamaindex_doc_store";
class PostgresDocumentStore extends _KVDocumentStore.KVDocumentStore {
    serializer = _utils.noneSerializer;
    constructor(config){
        const kvStore = new _PostgresKVStore.PostgresKVStore({
            schemaName: config?.schemaName,
            tableName: config?.tableName || DEFAULT_TABLE_NAME,
            ...config && "clientConfig" in config ? {
                clientConfig: config.clientConfig
            } : config && "client" in config ? {
                client: config.client,
                shouldConnect: config.shouldConnect ?? false
            } : {}
        });
        const namespace = config?.namespace || _global.DEFAULT_NAMESPACE;
        super(kvStore, namespace);
    }
}
