// @llamaindex/deepinfra/dist/index.cjs
Object.defineProperty(exports, '__esModule', { value: true });

var embeddings = require('@llamaindex/core/embeddings');
var utils = require('@llamaindex/core/utils');
var env = require('@llamaindex/env');
var openai = require('@llamaindex/openai');

const DEFAULT_MODEL$1 = "sentence-transformers/clip-ViT-B-32";
const API_TOKEN_ENV_VARIABLE_NAME = "DEEPINFRA_API_TOKEN";
const API_ROOT = "https://api.deepinfra.com/v1/inference";
const DEFAULT_TIMEOUT = 60 * 1000;
const DEFAULT_MAX_RETRIES = 5;
const mapPrefixWithInputs = (prefix, inputs)=>{
    return inputs.map((input)=>prefix ? `${prefix} ${input}` : input);
};
/**
 * DeepInfraEmbedding is an alias for DeepInfra that implements the BaseEmbedding interface.
 */ class DeepInfraEmbedding extends embeddings.BaseEmbedding {
    constructor(init){
        super(), this.getTextEmbeddings = async (texts)=>{
            const textsWithPrefix = mapPrefixWithInputs(this.textPrefix, texts);
            return this.getDeepInfraEmbedding(textsWithPrefix);
        };
        this.model = init?.model ?? DEFAULT_MODEL$1;
        this.apiToken = init?.apiToken ?? env.getEnv(API_TOKEN_ENV_VARIABLE_NAME) ?? "";
        this.queryPrefix = init?.queryPrefix ?? "";
        this.textPrefix = init?.textPrefix ?? "";
        this.maxRetries = init?.maxRetries ?? DEFAULT_MAX_RETRIES;
        this.timeout = init?.timeout ?? DEFAULT_TIMEOUT;
    }
    async getTextEmbedding(text) {
        const texts = mapPrefixWithInputs(this.textPrefix, [
            text
        ]);
        const embeddings = await this.getDeepInfraEmbedding(texts);
        return embeddings[0];
    }
    async getQueryEmbedding(query) {
        const text = utils.extractSingleText(query);
        if (text) {
            const queries = mapPrefixWithInputs(this.queryPrefix, [
                text
            ]);
            const embeddings = await this.getDeepInfraEmbedding(queries);
            return embeddings[0];
        } else {
            return null;
        }
    }
    async getQueryEmbeddings(queries) {
        const queriesWithPrefix = mapPrefixWithInputs(this.queryPrefix, queries);
        return await this.getDeepInfraEmbedding(queriesWithPrefix);
    }
    async getDeepInfraEmbedding(inputs) {
        const url = this.getUrl(this.model);
        for(let attempt = 0; attempt < this.maxRetries; attempt++){
            const controller = new AbortController();
            const id = setTimeout(()=>controller.abort(), this.timeout);
            try {
                const response = await fetch(url, {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                        Authorization: `Bearer ${this.apiToken}`
                    },
                    body: JSON.stringify({
                        inputs
                    }),
                    signal: controller.signal
                });
                if (!response.ok) {
                    throw new Error(`Request failed with status ${response.status}`);
                }
                const responseJson = await response.json();
                return responseJson.embeddings;
            } catch (error) {
                console.error(`Attempt ${attempt + 1} failed: ${error}`);
            } finally{
                clearTimeout(id);
            }
        }
        throw new Error("Exceeded maximum retries");
    }
    getUrl(model) {
        return `${API_ROOT}/${model}`;
    }
}

const ENV_VARIABLE_NAME = "DEEPINFRA_API_TOKEN";
const DEFAULT_MODEL = "mistralai/Mixtral-8x22B-Instruct-v0.1";
const BASE_URL = "https://api.deepinfra.com/v1/openai";
class DeepInfra extends openai.OpenAI {
    constructor(init){
        const { apiKey = env.getEnv(ENV_VARIABLE_NAME), additionalSessionOptions = {}, model = DEFAULT_MODEL, ...rest } = init ?? {};
        if (!apiKey) {
            throw new Error(`Set DeepInfra API key in ${ENV_VARIABLE_NAME} env variable`);
        }
        additionalSessionOptions.baseURL = additionalSessionOptions.baseURL ?? BASE_URL;
        super({
            apiKey,
            additionalSessionOptions,
            model,
            ...rest
        });
    }
}

exports.DeepInfra = DeepInfra;
exports.DeepInfraEmbedding = DeepInfraEmbedding;
