// llamaindex/dist/cjs/storage/indexStore/PostgresIndexStore.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PostgresIndexStore", {
    enumerable: true,
    get: function() {
        return PostgresIndexStore;
    }
});
const _global = require("@llamaindex/core/global");
const _PostgresKVStore = require("../kvStore/PostgresKVStore.js");
const _KVIndexStore = require("./KVIndexStore.js");
const DEFAULT_TABLE_NAME = "llamaindex_index_store";
class PostgresIndexStore extends _KVIndexStore.KVIndexStore {
    constructor(config){
        const kvStore = new _PostgresKVStore.PostgresKVStore({
            schemaName: config?.schemaName,
            tableName: config?.tableName || DEFAULT_TABLE_NAME,
            ...config && "clientConfig" in config ? {
                clientConfig: config.clientConfig
            } : config && "client" in config ? {
                client: config.client,
                shouldConnect: config.shouldConnect ?? false
            } : {}
        });
        const namespace = config?.namespace || _global.DEFAULT_NAMESPACE;
        super(kvStore, namespace);
    }
}
